subsys/rtio/rtio_executor.c

```diff
@@ -122,71 +125,108 @@ void rtio_executor_submit(struct rtio *r)
 /**
  * @brief Handle common logic when :c:macro:`RTIO_SQE_MULTISHOT` is set
  *
- * @param[in] r RTIO context
- * @param[in] curr Current IODev SQE that's being marked for finished.
- * @param[in] is_canceled Whether or not the SQE is canceled
+ * @param[in] iodev_sqe IODEV SQE that's being marked as finished.
+ * @param[in] result The result of the latest request iteration
+ * @param[in] is_ok Whether or not the SQE's result was successful
  */
-static inline void rtio_executor_handle_multishot(struct rtio *r, struct rtio_iodev_sqe *curr,
-						  bool is_canceled)
+static inline void rtio_executor_handle_multishot(struct rtio_iodev_sqe *iodev_sqe,
+						  int result, bool is_ok)
 {
-	/* Reset the mempool if needed */
-	if (curr->sqe.op == RTIO_OP_RX && FIELD_GET(RTIO_SQE_MEMPOOL_BUFFER, curr->sqe.flags)) {
-		if (is_canceled) {
-			/* Free the memory first since no CQE will be generated */
-			LOG_DBG("Releasing memory @%p size=%u", (void *)curr->sqe.rx.buf,
-				curr->sqe.rx.buf_len);
-			rtio_release_buffer(r, curr->sqe.rx.buf, curr->sqe.rx.buf_len);
-		}
+	if (!is_ok) {
+		/** Error handling of multi-shot submissions require
+		 * stopping re-submitting if something goes wrong.
+		 * Let the application decide what's best for handling
+		 * the corresponding error: whether re-submitting,
+		 * rebooting or anything else.
+		 */
+		iodev_sqe->sqe.flags |= RTIO_SQE_CANCELED;
+	}
+
+	struct rtio *r = iodev_sqe->r;
+	const bool is_canceled = FIELD_GET(RTIO_SQE_CANCELED, iodev_sqe->sqe.flags) == 1;
+	const bool uses_mempool = FIELD_GET(RTIO_SQE_MEMPOOL_BUFFER, iodev_sqe->sqe.flags) == 1;
+	const bool requires_response = FIELD_GET(RTIO_SQE_NO_RESPONSE, iodev_sqe->sqe.flags) == 0;
+	uint32_t cqe_flags = rtio_cqe_compute_flags(iodev_sqe);
+	void *userdata = iodev_sqe->sqe.userdata;
+
+	if (requires_response) {
+		rtio_cqe_submit(r, result, userdata, cqe_flags);
+	}
+
+	if (iodev_sqe->sqe.op == RTIO_OP_RX && uses_mempool) {
 		/* Reset the buffer info so the next request can get a new one */
-		curr->sqe.rx.buf = NULL;
-		curr->sqe.rx.buf_len = 0;
+		iodev_sqe->sqe.rx.buf = NULL;
+		iodev_sqe->sqe.rx.buf_len = 0;
 	}
-	if (!is_canceled) {
+
+	if (is_canceled) {
```

Comment:
is_cancelled || !is_ok and then you don’t need to mutate the sqe

Reply:
This branch should also likely occur before the cqe_submit

Reply:
Addressed.